### DISCUSSION TITLE
Ability to skip request using pre-request scripts

### DISCUSSION THREAD

**Discussion by brvaland** [Kong/insomnia#7628] [Help] [answered]
Hi Team

We are evaluating insomnia for our use-case and migration from postman and for few scenarios we folder created as "Regression" which we would like to run only in dev environment and would like to exclude in staging & production. 

In postman we used `postman.setNextRequest` and it worked fine but trying to find similar command or an alternative way to do same thing.

```
if (pm.environment.get("enable-regression") == "false") {
   postman.setNextRequest("request_name");
}
```
Please advise ?

Regards
Bhavesh

**Comment by subnetmarco** [comment]
@ihexxa do we support `setNextRequest` ?

**Comment by ihexxa** [comment]
Hello @brvaland, building request workflow with `setNextRequest` is not supported yet as it depends on "runner" to do this.
But we are already on it and hopefully it will be included in the next release, then you will be able to `run collection` or `run folder` in Insomnia.
BTW, will you trigger them with UI or CLI?

**Comment by brvaland** [comment]
@ihexxa we are planning to use both UI & CLI ? Currently we are using below folder structure.

```
- Customers [Folder]
   - Auth [Folder]
      - Get-Token [Request]
   - Regression [Folder]
      - < multiple folders and request >
   - Get Customers [Request]
```

In Get-Token Request [pre-request script] we use below code and based on environment we run all requests in regression folder or move to `Get Customers` request which will skip regression folder for staging and prod.

```
if (pm.environment.get("enable-regression") == "false") {
   postman.setNextRequest("Get Customers");
}
```

The option `run folder` will require us reorganise our collection and some changes to cicd to run via CLI. If we can support setNextRequest will make migration easy with no change.

**Comment by brvaland** [comment]
@subnetmarco @ihexxa please let me know if we can have both `run folder` and `insomnia.setNextReqest` support in 9.4 release ?

**Reply by subnetmarco** [reply]
It will ship between now and September. Not sure if in 9.4 or 9.5, but it's coming very soon.

**Reply by bpalm2006** [reply]
@subnetmarco @ihexxa any news on insomnia.execution.setNextRequest it would be awesome functionality that would allow repeating a request until it succeeds.

**Reply by ihexxa** [reply, marked as answer]
Hello @bpalm2006 @brvaland , the `setNextRequest` feature has been released together with the collection runner in v10.0.0.
You can find them in the snippets menu:
<img width="722" alt="Screenshot 2024-09-25 at 15 22 31" src="https://github.com/user-attachments/assets/a4f6d91f-b15f-4ac8-b1a0-f2c665936adc">
Also you can upload a csv file and which will be exposed as `insomnia.iterationData`.

**Reply by ihexxa** [reply]
@bpalm2006 I believe your case is not fully supported yet, will take a look at it.

**Reply by bpalm2006** [reply]
Yes @ihexxa, I'm running an After-response script on and endpoint where we are waiting for results and polling. I can see that setNextRequest seems to be a valid option but it still doesn't repeat the request.

Here is the script I'm using as an example (it is checking the body to see if the scores array has data and repeating until it does):

```
	var expectedHttpStatus = 200;
	var maxNumberOfTries = 10;
	var sleepBetweenTries = 5000;
	const jsonData = insomnia.response.json();

	if (!insomnia.environment.get("collection_tries")) {
	    insomnia.environment.set("collection_tries", 1);
	}

	if (((insomnia.response.code != expectedHttpStatus) || (jsonData.scores.length === 0)) && (insomnia.environment.get("collection_tries") < maxNumberOfTries)) {
		console.log("IN");
	    var tries = parseInt(insomnia.environment.get("collection_tries"), 10);
	    insomnia.environment.set("collection_tries", tries + 1);
	    setTimeout(function() {}, sleepBetweenTries);
	    insomnia.execution.setNextRequest(insomnia.info.requestName);
	} else {
		console.log("OUT");
	    insomnia.environment.unset("collection_tries");
		
		insomnia.test("Status code is " + expectedHttpStatus, function () {
				insomnia.response.to.have.status(expectedHttpStatus);
		});

		insomnia.test("Response time is less than 2000ms", function () {
				insomnia.expect(insomnia.response.responseTime).to.be.below(2000);
		});

		insomnia.test("Scores array should NOT be empty", function () {
				insomnia.expect(jsonData.scores).to.be.an('array').that.is.not.empty;
		});
	}
```

**Reply by ihexxa** [reply]
Starting a patch for it: https://github.com/Kong/insomnia/pull/8088

**Reply by ihexxa** [reply]
@bpalm2006 FYI the patch has been released with [10.1.1-beta.0](https://github.com/Kong/insomnia/releases/tag/core%4010.1.1-beta.0) and will be included in 10.1.1 in this week.

**Reply by bpalm2006** [reply]
Thank you!
